tests/src/psa_exercise_key.c

```diff
@@ -789,6 +789,9 @@ int mbedtls_test_psa_exported_key_sanity_check(
             TEST_EQUAL(1 + 2 * PSA_BITS_TO_BYTES(bits), exported_length);
             TEST_EQUAL(exported[0], 4);
         }
+    } else
+    if (PSA_KEY_TYPE_IS_DH_PUBLIC_KEY(type) || PSA_KEY_TYPE_IS_DH_KEY_PAIR(type)) {
+        TEST_LE_U(exported_length, PSA_BITS_TO_BYTES(bits));
```

Comment:
Can't we test for equality? I don't think we're removing leading zeros.

Reply:
I changed this to check equality. Now I know why I used less or equal check before. 
Now key generate tests fail randomly:
```
PASSED (5 / 5 tests (0 skipped))
przemek@przemek-VirtualBox:~/mbedtls/tests$ ./test_suite_psa_crypto
PSA generate key: FFDH, 2048 bits, good ........................... PASS
PSA generate key: FFDH, 3072 bits, good ........................... PASS
PSA generate key: FFDH, 4096 bits, good ........................... PASS
PSA generate key: FFDH, 6144 bits, good ........................... PASS
PSA generate key: FFDH, 8192 bits, good ........................... PASS

----------------------------------------------------------------------------

PASSED (5 / 5 tests (0 skipped))
przemek@przemek-VirtualBox:~/mbedtls/tests$ ./test_suite_psa_crypto
PSA generate key: FFDH, 2048 bits, good ........................... PASS
PSA generate key: FFDH, 3072 bits, good ........................... PASS
PSA generate key: FFDH, 4096 bits, good ........................... PASS
PSA generate key: FFDH, 6144 bits, good ........................... PASS
PSA generate key: FFDH, 8192 bits, good ........................... FAILED
  exported_length == PSA_EXPORT_PUBLIC_KEY_OUTPUT_SIZE(type, bits)
  at line 794, src/psa_exercise_key.c
  lhs = 0x00000000000003ff = 1023
  rhs = 0x0000000000000400 = 1024
```

It looks like generated key can have smaller length. The length of generated key is obtained using `mbedtls_mpi_size(&X)`. Is this correct?

Reply:
The key can have a leading zero byte, with very close to 1/256 probability. The PSA export format uses a constant size: a private key or a public key on the N-bit group is represented as an N-bit string. See the documentation of `psa_export_{public_,}key`.

Use `mbedtls_mpi_write_binary` with `buflen = PSA_BITS_TO_BYTES(bits)` (after checking that this value is less than the buffer size of course). `mbedtls_mpi_size` is not relevant here (it's only relevant when you want to strip leading zeros).

Reply:
```
psa_status_t mbedtls_psa_ffdh_generate_key(
    const psa_key_attributes_t *attributes,
    uint8_t *key_buffer, size_t key_buffer_size, size_t *key_buffer_length)
```

I'm using `mbedtls_mpi_write_binary` with `buflen = key_buffer_size`. This seems to be ok as I think `mbedtls_mpi_write_binary` fails when given buffer is too small.

`*key_buffer_length` is set to `mbedtls_mpi_size(&X);` and this should be changed to `PSA_BITS_TO_BYTES(attributes->core.bits)`?

Reply:
> I'm using mbedtls_mpi_write_binary with buflen = key_buffer_size

That's not correct if the output buffer is larger. Please make sure you have a test case for this.

(It might be a good idea for `import_export` and `import_export_public_key` to systematically test output with a larger buffer, but currently those need to be separate test cases.)

Reply:
Why larger `key_buffer_size` is a problem? We are returning actual key length via `key_buffer_length`.

Reply:
If you pass a larger size to `mbedtls_mpi_write_binary`, it pads with null bytes at the beginning (since the format is big-endian). E.g. for a 4-byte key and a 6-byte `output_buffer_size` the output would be 000012345678 (truncating to 00001234) instead of 12345678.

Reply:
Ok, but looking at the core part it validates kay bits against the alg and allocates slot for the key, so in the driver part the buffer is already validated (no need to check that `key_buffer_size == PSA_BITS_TO_BYTES(attributes->core.bits)`.

Reply:
Ah, ok, I hadn't realized that `key_buffer_size` was not the `output_size` argument passed to `psa_export_size`.